tests/algorithms/cca_tests.cpp

```diff
@@ -0,0 +1,170 @@
+/** TRACCC library, part of the ACTS project (R&D line)
+ *
+ * (c) 2021 CERN for the benefit of the ACTS project
+ *
+ * Mozilla Public License Version 2.0
+ */
+
+#include <gtest/gtest.h>
+
+#include <functional>
+#include <vecmem/memory/host_memory_resource.hpp>
+
+#include "clusterization/component_connection.hpp"
+#include "clusterization/measurement_creation.hpp"
+#include "definitions/primitives.hpp"
+#include "edm/cell.hpp"
+#include "edm/cluster.hpp"
+#include "tests/data_test.hpp"
+
+namespace {
+std::vector<std::string> names = {"hello", "goodbye"};
+
+struct cca_truth_hit {
+    uint64_t geometry_id = 0;
+    uint64_t hit_id = 0;
+    uint64_t num_cells = 0;
+    traccc::scalar channel0 = 0;
+    traccc::scalar channel1 = 0;
+    traccc::scalar variance0 = 0.;
+    traccc::scalar variance1 = 0.;
+
+    DFE_NAMEDTUPLE(cca_truth_hit, geometry_id, hit_id, num_cells, channel0,
+                   channel1, variance0, variance1);
+};
+
+using cca_truth_hit_reader = dfe::NamedTupleCsvReader<cca_truth_hit>;
+
+std::vector<std::string> get_test_files(void) {
+    const std::vector<std::pair<std::string, std::size_t>> cases = {
+        {"dense", 100},
+        {"multiple_module_single_hit", 100},
+        {"single_module_multiple_hit_single", 100},
+        {"single_module_multiple_hit_single_sparse", 100},
+        {"single_module_single_hit", 100},
+        {"very_dense", 100},
+        {"trackml_like", 30},
+    };
+    std::vector<std::string> out;
+
+    for (const std::pair<std::string, std::size_t> &c : cases) {
+        for (std::size_t i = 0; i < c.second; ++i) {
+            std::ostringstream ss;
+            ss << c.first << "_" << std::setfill('0') << std::setw(10) << i;
+            out.push_back(ss.str());
+        }
+    }
+
+    return out;
+}
+}  // namespace
+
+class ConnectedComponentAnalysisTests
+    : public traccc::tests::data_test,
+      public testing::WithParamInterface<std::string> {
+    public:
+    void test_connected_component_analysis(
```

Comment:
Why is this function separated from the `CpuSparseCCLAlgorithm` function? Helper functions like this mostly make sense if they are used from multiple places. But since this is only used in that one test, why not just implement one big function?

Reply:
The idea here is that these tests can be applied to any CCA algorithm, not just the CPU ones. That is why the design is so generic, and the idea of having this function be separate is that we can more easily apply it to, for example, a GPU implementation.